pykafka/producer.py

```diff
@@ -111,15 +125,114 @@ def __repr__(self):
             id_=hex(id(self))
         )
 
-    def _send_request(self, broker, req, attempt):
+    def __enter__(self):
+        """Context manager entry point - start the producer"""
+        self.start()
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        """Context manager exit point - stop the producer
+
+        If __exit__ is being called as a result of self._worker_exception
+        being raised, don't wait on the workers to finish - just re-raise
+        the exception. If __exit__ is being called because the with: block
+        has ended, do wait for workers to finish.
+        """
+        # If the thread crashed, don't wait for it
+        if self._worker_exception is not None:
+            log.exception(traceback)
+            self.stop(wait=False)
+            raise self._worker_exception
```

Comment:
I'm cheating and reading the context manager docs on the side, because I'd never written one, but those suggest this would be expected to `return False` instead of re-raising.